### DISCUSSION TITLE
spawn fore-exist project

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#40] [Ideas]
For 2-level validation a server-side component is necessary.

For validation and submission (dispatching to final target) there should be an OpenAPI endpoint:

* post `/submit?url=modules/somexql.xql`

To avoid duplicating validation processing fore-exist should generate the validation rules from the same html document the client processes.

There's been a prototypical implementation using generated XQueryX code but other solutions are equally welcome.

The process could be roughly:
1. client submits some instance data
2. the submission always targets the  url of the 'submission' endpoint of 'fore-exist' e.g. /fore-exist/submit?url=/myApp/modules/doSomething.xql'. 
3. If validation errors occur these are returned as a JSON structure
4. If validation succeeds the data are send along to the url given
5. eventually the app will want to return some data which need to be routed back to the client

**Comment by JoernT** [comment]
Alternatives:
* run server-side js and revalidate just the same as on client. How can that be done? Possible to wrap this whole thing up into a single eXist-db app? Of course this requires a node on the server. Is that a killer-argument?

* instead of compiled XQueryX use the slower 'evaluate' approach simply - iterating bind elements and doing an xquery evaluate. Though this is slower it's probably not severe as it only happens during final submit. 

* can't judge how much effort it would be to generate Schematron for the bindings

**Comment by windauer** [comment]
wrapped my head already around this and from my perspective I want to be able to use the exact same fore form with its calculations, bindings, validations on the server as well. From a trust point of view I don't trust the data nor the form but at least the form I can take from the server and use it to validate the client data. Usually the general "xforms" validation should be enough, but it would be great if I could have a validate-xsd function and a validate-schematron function, in both cases I'd like to either supply a full instance or a part of my instance, the 2nd argument could be the path to the xsd/schematron file.  Another possibility couls be to reference a xsd or schematron file on an instance.

**Reply by JoernT** [reply]
Yes, form is read from the Server . From that whatever validation generator generates the code to validate (no calculates needed here). However there is one problem: the validity of one node might depend on the validity of another. Means we got the gragh problem again that has been solved on the client already. This is a tricky one...

**Reply by JoernT** [reply]
Additional structural validators make sense and can be attached to the model. Typically the instance then declares which of the schemas to apply. This would be server-side directives.

**Comment by JoernT** [comment]
what about schematron? shouldn't it be possible to generate the schematron from the binds of a form?

**Comment by JoernT** [comment]
latest discussion with @DrRataplan revealed that the server don't even need to report on specific errors. If server validation fails it can only be the case cause data have been modified during request. It would be sufficient to report back that validation failed - the client already has the information to display the error. 

Instead the server can just return the failed data instance which the client can load as any usual loading of data and revalidate to show the alert messages.
